mllib/src/main/scala/org/apache/spark/ml/feature/HashingTF.scala

```diff
@@ -91,20 +90,13 @@ class HashingTF @Since("3.0.0") private[ml] (
   @Since("2.0.0")
   override def transform(dataset: Dataset[_]): DataFrame = {
     val outputSchema = transformSchema(dataset.schema)
-    val localNumFeatures = $(numFeatures)
-    val localBinary = $(binary)
+    val n = $(numFeatures)
+    val updateFunc = if ($(binary)) (v: Double) => 1.0 else (v: Double) => v + 1.0
 
     val hashUDF = udf { terms: Seq[_] =>
-      val termFrequencies = mutable.HashMap.empty[Int, Double].withDefaultValue(0.0)
-      terms.foreach { term =>
-        val i = indexOf(term)
-        if (localBinary) {
-          termFrequencies(i) = 1.0
-        } else {
-          termFrequencies(i) += 1.0
-        }
-      }
-      Vectors.sparse(localNumFeatures, termFrequencies.toSeq)
+      val map = new OpenHashMap[Int, Double]()
```

Comment:
This seems fine but is it faster than Scala's Map? the comment refers to the Java HashMap.

Reply:
yes, the comment in `util.collection.OpenHashMap` only refer to java hashmap, but to my knowledge, scala hashmap is slower than java hashmap.

and there is was also a [performance test](https://gist.github.com/rxin/44657c3f40d4e4294857) between java hashmap and scala hashmap: scala hashmap < java hashmap < scala openhashmap